client/src/includes/userbar.ts

```diff
@@ -0,0 +1,458 @@
+      '[data-wagtail-userbar-trigger]',
+    );
+    if (!trigger) return;
+    this.trigger = trigger;
+    const list = userbar.querySelector('[role=menu]');
+    if (!list) return;
+    const listItems = list.querySelectorAll('li');
+    const isActiveClass = 'is-active';
+
+    // Avoid Web Component FOUC while stylesheets are loading.
+    userbar.style.display = 'none';
+
+    /*
+    querySelector for all items that can be focused
+    tabIndex has been removed for roving tabindex compatibility
+    source: https://stackoverflow.com/questions/1599660/which-html-elements-can-receive-focus
+    */
+    const focusableItemSelector = `a[href],
+    button:not([disabled]),
+    input:not([disabled])`;
+
+    const showUserbar = (shouldFocus) => {
+      userbar.classList.add(isActiveClass);
+      trigger.setAttribute('aria-expanded', 'true');
+      // eslint-disable-next-line @typescript-eslint/no-use-before-define
+      list.addEventListener('click', sandboxClick, false);
+      // eslint-disable-next-line @typescript-eslint/no-use-before-define
+      window.addEventListener('click', clickOutside, false);
+
+      // Start handling keyboard input now that the userbar is open.
+      // eslint-disable-next-line @typescript-eslint/no-use-before-define
+      userbar.addEventListener('keydown', handleUserbarItemsKeyDown, false);
+
+      /*
+      The userbar has role=menu which means that the first link should be focused on popup
+      For weird reasons shifting focus only works after some amount of delay
+      Which is why we are forced to use setTimeout
+      */
+      if (shouldFocus) {
+        // Find the first focusable element (if any) and focus it
+        if (list.querySelector(focusableItemSelector)) {
+          setTimeout(() => {
+            // eslint-disable-next-line @typescript-eslint/no-use-before-define
+            setFocusToFirstItem();
+          }, 300); // Less than 300ms doesn't seem to work
+        }
+      }
+    };
+
+    const hideUserbar = () => {
+      userbar.classList.remove(isActiveClass);
+      trigger.setAttribute('aria-expanded', 'false');
+      // eslint-disable-next-line @typescript-eslint/no-use-before-define
+      list.addEventListener('click', sandboxClick, false);
+      // eslint-disable-next-line @typescript-eslint/no-use-before-define
+      window.removeEventListener('click', clickOutside, false);
+
+      // Cease handling keyboard input now that the userbar is closed.
+      // eslint-disable-next-line @typescript-eslint/no-use-before-define
+      userbar.removeEventListener('keydown', handleUserbarItemsKeyDown, false);
+    };
+
+    const toggleUserbar = (e2) => {
+      e2.stopPropagation();
+      if (userbar.classList.contains(isActiveClass)) {
+        hideUserbar();
+      } else {
+        showUserbar(true);
+      }
+    };
+
+    const isFocusOnItems = () =>
+      shadowRoot.activeElement &&
+      !!shadowRoot.activeElement.closest('.w-userbar-nav');
```

Comment:
Is `!!` JavaScript convention? An explicit check for null or whatever condition we're trying to avoid might be more readable.